### ISSUE TITLE
No indicator as to whether or not a synth recipe is a desynthesis.

### ISSUE THREAD

**Issue by Hozu** [DarkstarProject/darkstar#3544]
**_I have:_**

- [X] searched existing issues (http://github.com/darkstarproject/darkstar/issues/) to see if the issue I am posting has already been addressed or opened by another contributor
- [X] checked the commit log to see if my issue has been resolved since my server was last updated

**_Client Version_** (type `/ver` in game) **:**
current

**_Server Version_** (type `@revision` in game) **:**
b8a258559e8178b7924e43a95805e1d4992609bf

**_Source Branch_** (master/stable) **:**
master

**_Additional Information_** (Steps to reproduce/Expected behavior) **:**
When I was looking through synthutils.cpp for a recent pull request, I noticed that it's hardcoded to decrease the success rate for synths using a lightning element crystal. Not all such synths are desynths (ie. the alchemy ones to make Mercury from various fish), and not all desynths use a lightning element crystal (I can't think of one offhand). The synth_recipes table does not have a column for desynth. I did notice there is a column named Type, however it does not appear to even be read from judging by the query (if I'm reading it correctly). It looks like it has a value of either 0 or 1, though I didn't look at all of the entries closely.

**Comment by teschnei** [comment]
moblin armor (leathercraft) is a desynth that uses wind

**Comment by Deadwing888** [comment]
Quadav backplate desynth with a wind crystal

**Comment by TeoTwawki** [comment]
Quadav Backplates (and a ton of others) have multiple desynth recipes, 1 wind 1 lightning.

If that column is unused, lets rename it to make it obvious, and then use it to tell if a recipe is a desynth.

**Comment by wrenffxi** [comment]
I've QCed two craft recipe lists so far (alchemy and bonecraft) and the `type` column is consistently 1 for synths and 0 for desynths, so I believe we can reliably use this column to distinguish.

**Comment by maxtherabbit** [comment]
We will use it, but I'm almost positive at least some of the recipes in the
db are marked incorrectly. Appreciate you verifying them.

**Comment by ibm2431** [comment]
Should that column be renamed from `type` to something more informative?

**Comment by maxtherabbit** [comment]
I mean 'type' is a pretty accurate description for what it is, I think most
of the confusion around it rose from the fact it was never referenced in
the code

**Comment by ibm2431** [comment]
Just floating the idea of flipping the two values, and renaming the column from `type` to `is_desynth` ~!

**Comment by TeoTwawki** [comment]
I almost did something with that last night..I almost named it isDesynth while I was at it (no underscore, underscores in field names just feel wrong to me, dunno why they just do) but then decided to leave it alone. I didn't finish though. We should absolutely use 0 for normal synths and 1 for desynths, though we do not right now. All the older recipes have a 1. Some of the newer ones use zero. and all the desynths currently use zero.

@maxtherabbit the fact that the craft container is a recycled trade container broke my sleep deprived brain when I realized it when I was looking to try and store the field someplace instead of using a query every time you craft..wtf should I be doing with that. Or should I just leave it alone for now.

**Comment by maxtherabbit** [comment]
I'd leave it be IMO

**Comment by maxtherabbit** [comment]
"We should absolutely use 0 for normal synths and 1 for desynths" I just noticed this was backwards from how it is now... why would we reverse it? 1 seems to be a more intuitive value for a "constructive" synth and 0 for a "destructive" one

**Comment by TeoTwawki** [comment]
So it can be directly used as a bool when checking if its a desynth. We don't need to ask "is it a normal synth" when thats the majority situation, so we're going to be asking if its a desynth right? 1=true, 0=false. Since a bunch need changed *anyway* I'd rather just set them all in one go and use a bool. We'll never have a 3rd "type" there anyway.

**Comment by maxtherabbit** [comment]
I don't feel terribly strong about this either way but...

My opinion - using a more intuitive value for DB contributors is more 
valuable than saving the use of a single ! in the core

**Comment by TeoTwawki** [comment]
I don't see how zero is more intuitive, feels completely backward to me. it feels like someone will make the wrong assumption and exactly why I didn't want to do a ! or a == 0 to it. When I see it I am already thinking its an on/off for it being a desynth. /shrug

**Comment by TeoTwawki** [comment]
Maybe my way of thinking of it is because I am oriented on the "default" state. Oh well.

**Comment by maxtherabbit** [comment]
I think it's because in your mind you've already named the field `isDesynth` lol. If you think of it as `type` 1 feels like a more intuitive value for a synth that creates v. destroys, no?

**Comment by maxtherabbit** [comment]
if you guys think it is more intuitive to contributors to rename the field and invert the values, by all means do it - I'm really just discussing this because I find the differences in perception to be interesting

**Comment by TeoTwawki** [comment]
sort of? I think of it how I plan to use it. Which is basically the core asking the sql "is it a desynth?" And I think of positive "answers" as being 1 so..Circle of biased logic. 😄  never really thought of it in terms of create/destroy. *sploding magical crystals confuse me*

**Comment by Aethermage** [comment]
isDesynth makes sense because isSynth doesn't remove ambiguity any more than Type does.  

The reason you would use Synth to be 0 and Desynth to be 1 is because if we're going to use it as a flag, the flag has to refer to the name of the column. If the flag refers to a "Type" you now refer to the context of the data itself. We have two types of synthesis, one bearing the name of the entity itself that clearly comes first to mind when considering the two types. Also desynthesis wasn't even part of the original crafting concept to begin with. The only way you would consider the 0/1 false/true on/off dichotomy is if any of those made sense. They don't. It has entirely to do with what's more colloquial. The first type of synthesis is a regular synth, and the secondary type is a desynth. We don't start counting from one, we start counting from zero.

The only way around that is to name it isSynth, which technically they all are.

**Comment by takhlaq** [comment]
whatever just split them into their own tables for 0 ambiguity

**Comment by teschnei** [comment]
not going to swap the existing values - can you imagine if someone had started on a PR before and then didn't notice they were swapped while they were working on it?
to be honest, i had entertained the idea of using two tables instead of one, but i think it's more work than it's worth.  thus, the best path is just renaming "type" to something that makes more sense (and as aethermage pointed out, isSynth is out because it doesn't make any more sense)

**Comment by TeoTwawki** [comment]
> not going to swap the existing values - can you imagine if someone had started on a PR before and then didn't notice they were swapped while they were working on it?

The values we have aren't fully consistent, has to be fixed anyway. So doesn't matter one whit weather you use zero or one there, several PRs already merged with them at zero when the older entries were one. Yeah if only there was some kind of review process.. Oh wait. 👅

**Comment by TeoTwawki** [comment]
2 outa 3 devs so far want zero for desynth. I don't care enough let this keep going so I'll stfu now that I've said my piece.

**Comment by ibm2431** [comment]
"We made the decision last week to flip the values of type so they're more clear, could you run a quick regex on the rows you're adding to match up?"

This assumes that there is currently someone out there in the world working on a PR for this table, right now, which realistically speaking, is low odds. And on the chance there is, it'd take them all of two minutes to flip the values to match master.

And really, it's a bit silly to make design decisions because someone might be working with something we want to change. We'd never be able to fix something that's broken!

**Comment by teschnei** [comment]
No, the concern is that, if it slips by, it goes unnoticed because the mistake is completely valid data for this table.  This is very different from changing the API of something that would no longer compile, or the structure of the table that would make it no longer insert cleanly.

**Comment by maxtherabbit** [comment]
I think demo might be right with the separate table honestly

On Feb 16, 2017 7:11 PM, "Tyler Schneider" <notifications@github.com> wrote:

No, the concern is that, if it slips by, it goes unnoticed because the
mistake is completely valid data for this table. This is very different
from changing the API of something that would no longer compile, or the
structure of the table that would make it no longer insert cleanly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/DarkstarProject/darkstar/issues/3544#issuecomment-280507814>,
or mute the thread
